drv/sidecar-front-io/src/transceivers.rs

```diff
@@ -300,44 +301,148 @@ impl Transceivers {
-                            Addr::QSFP_CONTROL_PORT0 as u16 + u16::from(port),
-                            value,
-                        )?;
-                    }
-                }
+                let wdata: U16<byteorder::LittleEndian> = U16::new(mask.get());
+                fpga.write(op, addr, wdata)?;
             }
         }
         Ok(())
     }
 
+    /// Set power enable bits per the specified `mask`. Controls whether or not
+    /// a module's hot swap control will be turned on by the FPGA upon module
+    /// insertion.
+    pub fn enable_power<M: Into<FpgaPortMasks>>(
+        &self,
+        mask: M,
+    ) -> Result<(), FpgaError> {
+        self.masked_port_op(WriteOp::BitSet, mask, Addr::QSFP_POWER_EN0)
+    }
+
+    /// Clear power enable bits per the specified `mask`. Controls whether or
+    /// not a module's hot swap control will be turned on by the FPGA upon
+    /// module insertion.
+    pub fn disable_power<M: Into<FpgaPortMasks>>(
+        &self,
+        mask: M,
+    ) -> Result<(), FpgaError> {
+        self.masked_port_op(WriteOp::BitClear, mask, Addr::QSFP_POWER_EN0)
+    }
+
+    /// Set ResetL bits per the specified `mask`. This directly controls the
+    /// ResetL signal to the module.
+    pub fn deassert_reset<M: Into<FpgaPortMasks>>(
+        &self,
+        mask: M,
+    ) -> Result<(), FpgaError> {
+        self.masked_port_op(WriteOp::BitSet, mask, Addr::QSFP_MOD_RESETL0)
+    }
+
+    /// Clear ResetL bits per the specified `mask`. This directly controls the
+    /// ResetL signal to the module.
+    pub fn assert_reset<M: Into<FpgaPortMasks>>(
+        &self,
+        mask: M,
+    ) -> Result<(), FpgaError> {
+        self.masked_port_op(WriteOp::BitClear, mask, Addr::QSFP_MOD_RESETL0)
+    }
+
+    /// Set LpMode bits per the specified `mask`. This directly controls the
+    /// LpMode signal to the module.
+    pub fn assert_lpmode<M: Into<FpgaPortMasks>>(
+        &self,
+        mask: M,
+    ) -> Result<(), FpgaError> {
+        self.masked_port_op(WriteOp::BitSet, mask, Addr::QSFP_MOD_LPMODE0)
+    }
+
+    /// Clear LpMode bits per the specified `mask`. This directly controls the
+    /// LpMode signal to the module.
+    pub fn deassert_lpmode<M: Into<FpgaPortMasks>>(
+        &self,
+        mask: M,
+    ) -> Result<(), FpgaError> {
+        self.masked_port_op(WriteOp::BitClear, mask, Addr::QSFP_MOD_LPMODE0)
+    }
+
+    /// Have the SP execute a reset by manually writing the ResetL bits in the
+    /// FPGA. SFF-8679 states that the reset pulse needs to be 10 microseconds
+    /// (t_reset_init), but the module will not be ready for normal operation
+    /// until 2 seconds (t_reset) after reset is released.
+    pub fn module_reset<M: Into<FpgaPortMasks> + Copy>(
+        &self,
+        mask: M,
+    ) -> Result<(), FpgaError> {
+        self.assert_reset(mask)?;
+        userlib::hl::sleep_for(1);
+        self.deassert_reset(mask)?;
+        userlib::hl::sleep_for(2000);
```

Comment:
Hmmm, I'm not sure I love having this long wait on the Hubris side.

It makes it harder to misuse the API, but I'm worried about something else (e.g. network communication) hitting a timeout.

@bnaecker any opinions?  I'm wondering if we should put this on the host side instead.

Reply:
Doing that on the host side seems like a good idea. I know for sure that the Tofino SDE will do that manually, and we can always build a host-side method ourselves should we need it.

Reply:
I'm good with this as well. Admittedly I should have yanked this when we pulled out all the `PowerMode` stuff into the host, as those are higher order functions than the SP cares about anymore.

Given we now have `AssertReset` and `DeassertReset` messages, I'm going to remove the `Reset` message (the only caller of this function).